src/plugins/chatview/utils.js

```diff
@@ -71,3 +73,121 @@ export function resetElementHeight (ev) {
         ev.target.style = '';
     }
 }
+
+
+/**
+ * Handle XEP-0147 "query actions" invoked via xmpp: URIs.
+ * Supports message sending, roster management, and future actions.
+ *
+ * Example URIs:
+ *   xmpp:user@example.com?action=message&body=Hello
+ *   xmpp:user@example.com?action=add-roster&name=John&group=Friends
+ */
+export async function routeToQueryAction(event) {
+    const { u } = _converse.env;
+
+    try {
+        const uri = extractXMPPURI(event);
+        if (!uri) return;
+
+        const { jid, queryParams } = parseXMPPURI(uri);
+        if (!u.isValidJID(jid)) {
+            return log.warn(`Invalid JID: "${jid}"`);
+        }
+
+        const action = queryParams.get('action');
+        if (!action) {
+            log.debug(`No action specified, opening chat for "${jid}"`);
+            return api.chats.open(jid);
+        }
+
+        switch (action) {
+            case 'message':
+                await handleMessageAction(jid, queryParams);
+                break;
+
+            case 'add-roster':
+                await handleRosterAction(jid, queryParams);
+                break;
+
+            default:
+                log.warn(`Unsupported XEP-0147 action: "${action}"`);
+                await api.chats.open(jid);
+        }
+    } catch (error) {
+        log.error('Failed to process XMPP query action:', error);
+    }
+}
+
+/**
+ * Extracts and decodes the xmpp: URI from the window location or hash.
+ */
+function extractXMPPURI(event) {
+    let uri = null;
+
+    // Case 1: protocol handler (?uri=...)
+    const searchParams = new URLSearchParams(window.location.search);
+    uri = searchParams.get('uri');
+
+    // Case 2: hash-based (#converse/action?uri=...)
+    if (!uri && location.hash.startsWith('#converse/action?uri=')) {
+        event?.preventDefault();
+        uri = location.hash.split('uri=').pop();
+    }
+
+    if (!uri) return null;
+
+    // Decode URI and remove xmpp: prefix
+    uri = decodeURIComponent(uri);
+    if (uri.startsWith('xmpp:')) uri = uri.slice(5);
+
+    // Clean up URL (remove ?uri=... for a clean view)
+    const cleanUrl = `${window.location.origin}${window.location.pathname}`;
+    window.history.replaceState({}, document.title, cleanUrl);
+
+    return uri;
+}
+
+/**
+ * Splits an xmpp: URI into a JID and query parameters.
+ */
+function parseXMPPURI(uri) {
+    const [jid, query] = uri.split('?');
+    const queryParams = new URLSearchParams(query);
+    return { jid, queryParams };
+}
+
+/**
+ * Handles the `action=message` case.
+ */
+async function handleMessageAction(jid, params) {
+    const body = params.get('body') || '';
+    const chat = await api.chats.open(jid);
+
+    if (body && chat) {
+        await chat.sendMessage({ body });
+    }
+}
+
+/**
+ * Handles the `action=add-roster` case.
+ */
+async function handleRosterAction(jid, params) {
+    await api.waitUntil('connected');
+    await api.waitUntil('rosterContactsFetched');
```

Comment:
I think you only need the 2nd `waitUntil` because by the time `rosterContactsFetched` has resolved, `connected` will also have resolved. Or is there a specific reason you have both?